deploy/examples/create-external-cluster-resources.py

```diff
@@ -1134,7 +1142,24 @@ def create_rgw_admin_ops_user(self):
                     % (cmd, execErr.output, execErr.returncode, execErr.stderr)
                 )
                 raise Exception(err_msg)
-
+            if self._arg_parser.rgw_realm_name:
```

Comment:
Sorry I don't understand this part why to do need to modify realms for existing rados-admin-ops user?

Reply:
@thotz ^^

Reply:
We only need to worry about the existing users and their realms. If `realm` is not passed then it is okay to assume there is no realm in the ceph RGW cluster. 

I can understand point, in the old ODF cluster user may be created without realm and with this option realm can be specified etc. If the realm is not specified then user always gets created in the default realm(if it exists). 

Consider below example, I have two realms --> realm-a and realm-b.
two zonegroups --> zone-group-a (configured with realm-a), zone-group-b(configured with realm-b)
two zones --> zone-a belongs to zone-group-a, zone-b belongs to zone-group-b
two rgw-servers -> rgw-a belongs to zone-a, rgw-b belongs to zone-b.

Let us assume realm-a is default.

If the user tries to import rgw-a with help of script, all the operations will work. because adminops user will be created in realm-a. rgw-a server can be accessed via ceph rgw users in realm-a

If the user tries to import rgw-b with help of script, ideally this should not work, Rook won't able to communicate with rgw-b with adminops created in realm-a. If it is exporting the rgw-b server and then all the follow up operation   is falling like OBC/cephobjectstoreuser creation,deletion etc.

Otherwise there is no need to update the realm for already working adminops-user for old cluster

Reply:
`Let us assume realm-a is default.`

But what if the user wants to change the realm with time, probably that's the reason ceph has `--realm-new-name` this flag.

Reply:
there would be only a single user created with the name `self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME` so if the script would run on multiple cluster with multiple realm the user needs to be updated
@thotz

Reply:
For that we have the realm argument, I still don't understand modifying realm for existing adminops user
If the earlier user was created without realm, then IMO better to create new user in the realm than modify the existing user to new realm

Reply:
CC @cbodley @skoduri @alimaredia 

for creating user in multisite only --rgw-realm is required right, do zone/zonegroup is also a must?

What are chances of having multiple realms in the multisite configuration, is it common use case??

Reply:
> for creating user in multisite only --rgw-realm is required right, do zone/zonegroup is also a must?

`radosgw-admin` commands always execute in a specific zone. if you omit the `--rgw-zonegroup` and `--rgw-zone` arguments, it will try to use the realm's default zone/zonegroup. if you haven't set those defaults (either with the `--default` option to `radosgw-admin zone/zonegroup create/modify`, or with `radosgw-admin zone/zonegroup default`), the command should fail

since rook may be managing many realms/zonegroups/zones, it shouldn't treat any as special so should avoid relying on these defaults. rook should be specifying all of `--rgw-realm` and `--rgw-zonegroup` and `--rgw-zone` for every admin command it runs

for `radosgw-admin user create`, you need to run it on the master zone of the master zonegroup. it will replicate to other zones in the realm from there

> What are chances of having multiple realms in the multisite configuration, is it common use case??

think of each realm as a separate multisite configuration. i expect it would be very common for rook users to create several unrelated ObjectStores, which would each get their own unique zone/zonegroup/realm

multisite configuration would only be necessary when they want to replicate between these ObjectStores

Reply:
> > for creating user in multisite only --rgw-realm is required right, do zone/zonegroup is also a must?
> 
> `radosgw-admin` commands always execute in a specific zone. if you omit the `--rgw-zonegroup` and `--rgw-zone` arguments, it will try to use the realm's default zone/zonegroup. if you haven't set those defaults (either with the `--default` option to `radosgw-admin zone/zonegroup create/modify`, or with `radosgw-admin zone/zonegroup default`), the command should fail
> 

@cbodley The helper script is used for importing external ceph clusters and is executed separately. Rook is not involved here. Only `radosgw-admin` command in this script executes the user creation. In this case, if even --rgw-zone/--rgw-zonegroup is not passed the command will pass if default zone/zonegroup is set on the cluster. Otherwise, it will fail. And I am assuming default zone/zonegroup does not refers to master zone/zonegroup in the realm. 
So based on the above comment, do you think is it better to include rgw-zone / rgw-zonegroup for this helper script in external ceph cluster usecases? Please note this script does not pass any info to Rook Operator. It is just creates resources required for importing external cluster for Rook. Users can manually create the resources even without using this script



> since rook may be managing many realms/zonegroups/zones, it shouldn't treat any as special so should avoid relying on these defaults. rook should be specifying all of `--rgw-realm` and `--rgw-zonegroup` and `--rgw-zone` for every admin command it runs
> 
> for `radosgw-admin user create`, you need to run it on the master zone of the master zonegroup. it will replicate to other zones in the realm from there
> 
> > What are chances of having multiple realms in the multisite configuration, is it common use case??
> 
> think of each realm as a separate multisite configuration. i expect it would be very common for rook users to create several unrelated ObjectStores, which would each get their own unique zone/zonegroup/realm
> 
> multisite configuration would only be necessary when they want to replicate between these ObjectStores

Reply:
> And I am assuming default zone/zonegroup does not refers to master zone/zonegroup in the realm.

correct. multisite was designed for disaster recovery, where you'd generally only put one zone on each ceph cluster. so the primary zone would be the default on its cluster, and the secondary zone would be the default on the other cluster. in that scenario, the defaults make things simpler to manage. but whenever you have multiple zones or zonegroups on a cluster, defaults are likely to confuse things

> do you think is it better to include rgw-zone / rgw-zonegroup for this helper script in external ceph cluster usecases?

i think the script will work more reliably if you require the zone and zonegroup options

Reply:
Sure @parth-gr